### ISSUE TITLE
document.offsetAt appears to return incorrect values in onDidChangeTextDocument when multiple edits made

### ISSUE THREAD

**Issue by DanTup** [Microsoft/vscode#10047]
- VSCode Version: 1.3.1
- OS Version: Windows 10

I'm developing an extension which needs to send content changes to a separate service that is doing analysis. I'm using the `onDidChangeTextDocument` event.

Everything works fine if I'm only typing with one cursor; however if I use `Alt` to place multiple cursors, then I get the wrong offsets from `offsetAt` (presumably because it's using the new state of the document, not the one that the `Position` supplies was based from).

Example:

Create a file that contains two commented lines and no whitespace, but CrLfs:

    //
    //

Place your cursor on the end of the second line and type a character. You'll get an event saying the edit is on Line 1, Character 2 (this is correct, it's zero-based):

    {"range":[{"line":1,"character":2},{"line":1,"character":2}],"rangeLength":0,"text":"a"},

If you feed this into document.offsetAt you will get back 6 (this is also correct; you have `SLASH` `SLASH` `CR` `LR` `SLASH` `SLASH` proceeding it:

    e.document.offsetAt(e.contentChanges[0].range.start) // 6

Now do the same thing, but with multiple edit cursors - one at the end of each line. You'll get an event for both lines at Character 2 (again, correct):

    {"range":[{"line":1,"character":2},{"line":1,"character":2}],"rangeLength":0,"text":"a"},
    {"range":[{"line":0,"character":2},{"line":0,"character":2}],"rangeLength":0,"text":"a"}

However, if you feed these into `offsetAt`, the first one (which is the change lower down the document) will return 7!

    e.document.offsetAt(e.contentChanges[0].range.start) // 7

My guess is that it's because `offsetAt` is now seeing the `a` inserted on the first line, shifting the calculation.

Possible fix might be to provide the original document along with a `TextDocumentChangeEvent`?

**Comment by alexandrudima** [comment]
Indeed, the text document is already updated and the conversion methods operate on its new contents.

We can think about including offset information in the events

**Comment by DanTup** [comment]
Is it likely this will be added? I have some hacks to workaround it, but if it's likely to remain this way I think it's worth me tidying them up.

**Comment by DanTup** [comment]
@alexandrudima This really feels like a bug - there's no simple workaround that allows me to know what part of the document was modified. Is it complicated to fix? :(

**Comment by DanTup** [comment]
@alexandrudima Awesome, thank you! I'll test it out in insiders when it lands :)

**Comment by DanTup** [comment]
LGGTM in Insiders, thanks! 👍
